Add async compute performance sample

HansKristian-Work opened the pull request:
## Description

This sample demonstrates how multiple queues should be used on TBDR (Arm) architectures to keep all queues busy with work. The sample uses a compute shader post-processing pipeline, which normally is a problem on mobile, but it can be made to work well if we use multiple queues to remove starvation of fragment hardware queue.

Also demonstrates more direct use of semaphores, queue ownership transfers.

Tested on Arm, RADV and NV.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

HansKristian-Work committed:
Set acquired semaphore directly in begin_frame().

Makes it possible for samples to call it directly.
Also, be robust against multiple RenderContext::begin() calls.

HansKristian-Work committed:
Don't push constants if pipeline layout doesn't use it.

HansKristian-Work committed:
Support overriding rasterization state in vkb::GeometrySubpass.

Required for depth bias overrides.

HansKristian-Work committed:
Support queue ownership transfer in image_memory_barrier.

HansKristian-Work committed:
Make submission logic more flexible.

HansKristian-Work committed:
Add support for enabling high priority graphics queue.

Used by Async compute sample to get desired pre-empt behavior.

HansKristian-Work committed:
Don't blindly assume where WSI acquire semaphore is consumed.

A frame context is created after we request the semaphore, so we need to
keep ownership until it's released.

HansKristian-Work committed:
Add async compute sample.

HansKristian-Work committed:
Fix broken hashing of SPIR-V.

We would only hash the lower 8 bits of every SPIR-V word ...
Triggered collision in async compute sample.

HansKristian-Work committed:
Add tutorial.

HansKristian-Work committed:
Add missing samples README.md entry.

HansKristian-Work committed:
Workaround Doxygen CI error.

HansKristian-Work committed:
Update copyrights.

HansKristian-Work committed:
Fix clang-format.

HansKristian-Work committed:
Update another copyright check.

SaschaWillems commented on framework/core/device.cpp:
```diff
@@ -40,7 +40,27 @@ Device::Device(PhysicalDevice &gpu, VkSurfaceKHR surface, std::unordered_map<con
+			uint32_t graphics_queue_family = get_queue_family_index(VK_QUEUE_GRAPHICS_BIT);
+			if (graphics_queue_family == queue_family_index)
+			{
+				queue_priorities[queue_family_index].reserve(queue_family_property.queueCount);
```
I think this should be ```resize``` instead of ```reserve```. With the latter, this crashes for my on Windows with MSVC2019.

HansKristian-Work replied:
Fixed.

SaschaWillems reviewed (changes requested):
This is a great sample, as usual 👍 

I noticed a few problems tough, see bewlo.

SaschaWillems commented on framework/rendering/render_context.cpp:
```diff
@@ -409,6 +415,19 @@ void RenderContext::end_frame(VkSemaphore semaphore)
+
+VkSemaphore RenderContext::consume_acquired_semaphore()
+{
+	assert(frame_active && "Frame is not active, please call begin_frame");
```
This error is triggered on some of the existing samples for me:

![image](https://user-images.githubusercontent.com/10283127/114220266-0e473f00-996c-11eb-8253-a1b5c4615f9d.png)

Tested with the MSAA and afbc samples, that shows above error in debug mode after the first few frames and then exists.

HansKristian-Work replied:
Fixed.

SaschaWillems reviewed (changes requested):
A change in the framework causes some of the existing samples to crash for me.

HansKristian-Work committed:
Fix silly vector bug.

HansKristian-Work committed:
Release frame after releasing semaphore.

gary-sweet commented on samples/performance/async_compute/async_compute.cpp:
```diff
@@ -0,0 +1,809 @@
+
+void AsyncComputeSample::prepare_render_targets()
+{
+	// To make this sample demanding enough to saturate mobile devices, use 4K.
```
Using a size here that is larger than the swapchain causes significant performance issues on some of our devices (which are designed to be very small and low power). You noted that this was to allow comparing of cross-device performance, but I don't believe that's the aim of these performance samples; they allow you to compare various options on a single device.

HansKristian-Work replied:
I hit vsync on the devices I test on unless I really push it. The intention is to be able to see a difference when toggling the options, not run fast on all devices. (The sample is tagged Arm as well, FWIW).

SaschaWillems replied:
Samples based on the ARM framework are always locked to vsync. We discussed this on the last call and hopefully have fix for that soon ;)

gary-sweet replied:
> The sample is tagged Arm as well, FWIW

I'm not sure why it's tagged Arm. What here is Arm specific?

HansKristian-Work replied:
It was tagged like that since it's designed to demonstrate performance differences on Mali. It's somewhat implementation specific how this works (how many queues do you have, can queues pre-empt, is it TBDR, etc).

gary-sweet replied:
Sure, but that's true of many of the performance samples, which aren't tagged as Arm specific.
My understanding was that the tag was for things that really are vendor specific.

HansKristian-Work replied:
Fixed.

gary-sweet replied:
Thanks.

gary-sweet commented on samples/performance/async_compute/async_compute.cpp:
```diff
@@ -0,0 +1,809 @@
+	};
+
+	// 8K shadow-map overkill to stress devices.
+	vkb::core::Image shadow_target{*device, VkExtent3D{8 * 1024, 8 * 1024, 1}, VK_FORMAT_D16_UNORM,
```
8K images aren't universally supported by Vulkan (or some of our devices).  You should be using 4K maximum for guaranteed support, or checking the limits and selecting the best-fit power-of-two size perhaps.

HansKristian-Work replied:
Fixed.

gary-sweet replied:
Looks good, thanks.

gary-sweet reviewed (commented):
Just a couple of comments after a brief test.  Unfortunately on some of our devices, this breaks some of the limits and runs very slowly.

HansKristian-Work committed:
Clamp shadow map size to device limits.

4K is minspec in Vulkan, and not all mobile devices support 8K render
targets.

gary-sweet commented on samples/performance/async_compute/async_compute.cpp:
```diff
@@ -0,0 +1,820 @@
+	    vkb::StatIndex::gpu_cycles,
+	    vkb::StatIndex::gpu_vertex_cycles,
+	    vkb::StatIndex::gpu_fragment_cycles,
+	});
```
I think you need to add "vkb::CounterSamplingConfig{vkb::CounterSamplingMode::Continuous}" as a second argument to request_stats here. That will prevent the Vulkan extension for performance counters being used.

The problem with the Vulkan counters extension is that it is designed to operate on a single command buffer (it's a query pool thing). This sample uses at least four different command buffers, each of which is starting and stopping stats sampling, and the Vulkan counters then just don't work.

You can then remove all the begin_sampling() and end_sampling() calls.

HansKristian-Work replied:
Ye, I guess I can fix that.

HansKristian-Work replied:
Continous seems to be broken. The output is very jumpy and weird, and I don't know why. I removed begin/end_sampling which should work. I cannot reproduce any issues here. I just pushed a commit which removes the begin/end_sampling calls.

gary-sweet replied:
The pipeline_barriers sample uses continuous sampling. Does that work correctly?
We don't support continuous sampling here, someone at Arm would need to look into that.

gary-sweet replied:
> I removed begin/end_sampling which should work.

It does, but you'll also need to remove the stats that are listed. Right now, they appear in the UI but are always 0, since nothing is updating them any more.  This is why continuous sampling mode would be better.  The counters won't display when continuous mode isn't supported then, but would still work when they can.

HansKristian-Work replied:
Found some bugs in the Continuous implementation. Fixed those, and flipped the sample over to continuous.

gary-sweet replied:
Thanks a lot. That looks much better.

gary-sweet commented on samples/performance/async_compute/async_compute.cpp:
```diff
@@ -0,0 +1,820 @@
+	auto &queue          = *early_graphics_queue;
+	auto &command_buffer = render_context->get_active_frame().request_command_buffer(queue);
+	command_buffer.begin(VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT);
+	stats->begin_sampling(command_buffer);
```
Here for example

HansKristian-Work committed:
Remove begin/end_sampling().

Won't work with multiple command buffers per frame in current framework.

HansKristian-Work committed:
Use continuous sampling mode.

HansKristian-Work committed:
Fix Continuous sampling.

- Don't push samples from thread when we're not supposed to.
  Fixes latency issues where it takes a very long time to observe new
  counters.
- Keep track of fractional sample count to keep a steady pace.
- Fix formula for sample_count. Was multiplying by the wrong value, and
  rounding too early.

HansKristian-Work committed:
Fix style.

HansKristian-Work committed:
Remove vendor specific tag.

gary-sweet commented on samples/performance/async_compute/async_compute.cpp:
```diff
@@ -0,0 +1,815 @@
+
+	// We can potentially get more overlap if we double buffer the HDR render target.
+	// In this scenario, the next frame can run ahead a little further before it needs to block.
+	if (double_buffer_hdr_frames)
```
I think you need an 'else' here to reset forward_render_target_index to 0 when switching out of double buffer mode.
Double-buffer mode (for me) gives a 'lumpy' framerate (faster then slower every second or so). I noticed that when switching double-buffer mode off, the lumpiness continued.  Adding the 'else' case fixes the issue for me.

HansKristian-Work replied:
Not exactly sure why, but it's somewhat cleaner.

HansKristian-Work committed:
Clear frame index to 0 when not doing double buffer HDR.

gary-sweet reviewed (approved):
Thanks for the fixes.

SaschaWillems reviewed (approved):
LGTM. Works fine for me on Win 10 + NV RTX 2060 and my Google Pixel 4a with an Adreno 618.

HansKristian-Work commented:
@marty-johnson59 Good to go.
